include/openthread/tcp.h

```diff
@@ -0,0 +1,616 @@
+    const uint8_t *mData;          ///< Pointer to data referenced by this linked buffer
+    uint16_t mLength;              ///< Length of this linked buffer
+} otLinkedBuffer;
+
+struct otTcpEndpoint;
+typedef struct otTcpEndpoint otTcpEndpoint;
+
+/**
+ * This callback informs the application that the TCP 3-way handshake is
+ * complete and that the connection is now established.
+ */
+typedef void (*otTcpEstablished)(void *aContext, otTcpEndpoint *aEndpoint);
+
+/**
+ * This callback informs the application that data in the provided
+ * @p otLinkedBuffers have been acknowledged by the connection peer and that
+ * the relevant @p otLinkedBuffers and the data they contain can be reclaimed
+ * by the application. The @p otLinkedBuffers are guaranteed to be identical to
+ * those passed in to TCP via @p otTcpSendByReference.
+ */
+typedef void (*otTcpSendDone)(void *aContext, otTcpEndpoint *aEndpoint, otLinkedBuffer *aData);
+
+/**
+ * This callback informs that the first @p aNumBytes in the send buffer have
+ * been acknowledged by the connection peer and that their underlying memory
+ * can be reclaimed by the application.
+ *
+ * This callback is not necessary for correct TCP operation. Most applications
+ * can just rely on the @p otTcpSendDone callback. If an application wants to
+ * be able to receive fine-grained feedback as memory in the send buffer
+ * becomes available again (instead of waiting for an entire linked buffer's
+ * worth of data becomes available), it can register this callback.
+ */
+typedef void (*otTcpBytesAcked)(void *aContext, otTcpEndpoint *aEndpoint, size_t aNumBytes);
+
+/**
+ * This callback informs the application that if data is added to the send
+ * buffer, some of it will be transmitted immediately without delay, as opposed
+ * to being queued for transmission once the peer ACKs some data. After a call
+ * to @p otTcpSendByReference, an @p otTcpSendReady callback is guaranteed to
+ * be called, either immediately (if the connection is already ready) or
+ * sometime in the future (once the connection becomes ready for more data).
+ *
+ * This callback is not necessary for correct TCP operation. If more data is
+ * added to the send buffer than can be transmitted without delay, it will
+ * simply be queued for transmission at a later time. This callback should be
+ * used only in cases where some assurance is desired that data added to the
+ * send buffer will be sent soon (e.g., TCP won't wait for the recipient to
+ * ACK some other data first before sending this data out). For example, you
+ * may use this callback if you'd rather have your data be dropped than develop
+ * a backlog of data in your send buffer. But for most applications, where this
+ * isn't a concern, it's expected that one would not use this callback at all.
+ */
+typedef void (*otTcpSendReady)(void *aContext, otTcpEndpoint *aEndpoint);
+
+/**
+ * This callback indicates the number of bytes available for consumption from
```

Comment:
nit: have brief in separate block above

Reply:
Done. I'm not 100% what you meant by "in separate block"; can you please check that what I've done is acceptable?